src/orion/core/worker/multi_task_algo.py

```diff
@@ -0,0 +1,551 @@
+        By default, the cardinality of the specified search space will be used to check
+        if all possible sets of parameters has been tried.
+        """
+        if not self._enabled:
+            return self.algorithm.is_done
+        # NOTE: DEBUGGING:
+        total_trials = len(self.unwrapped._trials_info)
+        trials_from_target_task = sum(
+            self.get_task_id(point) == 0
+            for point, result in self.unwrapped._trials_info.values()
+        )
+        trials_from_other_tasks = total_trials - trials_from_target_task
+        log.debug(
+            f"Trials from target task: {trials_from_target_task}, "
+            f"trials from other tasks: {trials_from_other_tasks} "
+        )
+        log.debug(f"self.n_observed: {self.n_observed}")
+        log.debug(f"self.n_observed: {self.n_suggested}")
+        log.debug(f"wrapped algo.is_done: {self.algorithm.is_done}")
+
+        if trials_from_target_task >= self.space.cardinality:
+            return True
+
+        max_trials = getattr(self, "max_trials", float("inf"))
+        if trials_from_target_task >= max_trials:
+            return True
+
+        return False
+
+    @contextmanager
+    def in_task(self, task_id: int):
+        """ Contextmanager that temporarily changes the value of `self.current_task_id`
+        to `task_id` and restores the original value after exiting the with block.
+        """
+        previous_task_id = self.current_task_id
+        self.current_task_id = task_id
+        yield
+        self.current_task_id = previous_task_id
+
+    @singledispatchmethod
+    def _add_task_id(self, point: Any, task_id: int) -> Any:
+        raise NotImplementedError(point)
+
+    @_add_task_id.register(np.ndarray)
+    @_add_task_id.register(tuple)
+    def _add_task_id_to_point(self, point: Any, task_id: int) -> Tuple[int, Any]:
+        # NOTE: Need to do a bit of gymnastics here because the points are expected
+        # to be tuples. In order to add the task id at the right index, we convert
+        # them like so:
+        # Tuple -> Trial -> dict (.params) -> dict (add task_id) -> Trial -> Tuple.
+        trial = tuple_to_trial(point, self.space)
+        trial_with_task_id = self._add_task_id(trial, task_id)
+        point_with_task_id = trial_to_tuple(trial_with_task_id, self.algorithm.space)
+        return point_with_task_id
+
+    @_add_task_id.register
+    def _add_task_id_to_trial(self, trial: Trial, task_id: int) -> Trial:
+        # NOTE: Need to do a bit of gymnastics here because the points are expected
+        # to be tuples. In order to add the task id at the right index, we convert
+        # them like so:
+        # Trial -> dict (.params) -> dict (add task_id) -> Trial.
+        params = trial.params.copy()
+        params["task_id"] = task_id
+        # TODO: Should we copy over the status and everything else?
+        trial_with_task_id = dict_to_trial(params, self.algorithm.space)
```

Comment:
But we lose all other info in the task while doing so. An alternative would be

```python
trial_with_task_id = copy.deepcopy(trial)
task_id_index = list(space.keys()).index('task_id')
trial_with_task_id._params.insert(task_id_index, Trial.Param(name='task_id', value=value, type='categorical'))
```

Note that this operation is done on the original trial, without any transformation. Transformations should be applied after the task id is added (somewhere outside of _add_task_id_to_trial).